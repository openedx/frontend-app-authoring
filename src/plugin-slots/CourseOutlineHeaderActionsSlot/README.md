# CourseOutlineHeaderActionsSlot

### Slot ID: `course_outline_header_actions_slot`
### Plugin Props:

* `isReIndexShow` - Boolean.
* `isSectionsExpanded` - Boolean.
* `headerNavigationsActions` - Object. See Proptypes definition in the component for details.
* `hasSections` - Boolean. Indicates if the course outline has sections.
* `courseActions` - Object. See Proptypes definition in the component for details.
* `errors` - Object. See Proptypes definition in the component for details.
* `sections` - Array of objects. Sections of the course outline.

## Description

The slot is positioned in the `SubHeader` section of the Course Outline page, suitable for adding action buttons.

The slot by default contains the action buttons such as `+ New Section`, `Reindex`, `View Live`.

## Example

![Screenshot of a custom button in slot](./images/mybutton_added_to_the_slot.png)

The following example configuration inserts an extra button to the header as shown above.

```js
import { DIRECT_PLUGIN, PLUGIN_OPERATIONS } from '@openedx/frontend-plugin-framework';
import { Button } from '@openedx/paragon';

const MyButton = () => (
  <Button>üê£</Button>
);

const config = {
  pluginSlots: {
    course_outline_header_actions_slot: {
      keepDefault: true,
      plugins: [
        {
          op: PLUGIN_OPERATIONS.Insert,
          widget: {
            id: 'my-extra-button',
            priority: 60,
            type: DIRECT_PLUGIN,
            RenderWidget: MyButton,
          },
        },
      ]
    }
  },
}

export default config;
```

