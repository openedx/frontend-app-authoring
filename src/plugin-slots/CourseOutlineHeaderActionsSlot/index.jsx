import React from 'react';
import PropTypes from 'prop-types';
import { PluginSlot } from '@openedx/frontend-plugin-framework';

import HeaderNavigations from 'CourseAuthoring/course-outline/header-navigations/HeaderNavigations';

const CourseOutlineHeaderActionsSlot = ({
  headerNavigationsActions,
  isReIndexShow,
  isSectionsExpanded,
  isDisabledReindexButton,
  hasSections,
  courseActions,
  errors,
  sections,
}) => (
  <PluginSlot
    id="course_outline_header_actions_slot"
    pluginProps={{
      isReIndexShow,
      isSectionsExpanded,
      isDisabledReindexButton,
      headerNavigationsActions,
      hasSections,
      courseActions,
      errors,
      sections,
    }}
  >
    <HeaderNavigations
      headerNavigationsActions={headerNavigationsActions}
      isReIndexShow={isReIndexShow}
      isSectionsExpanded={isSectionsExpanded}
      isDisabledReindexButton={isDisabledReindexButton}
      hasSections={hasSections}
      courseActions={courseActions}
      errors={errors}
    />
  </PluginSlot>
);

CourseOutlineHeaderActionsSlot.propTypes = {
  isReIndexShow: PropTypes.bool.isRequired,
  isSectionsExpanded: PropTypes.bool.isRequired,
  isDisabledReindexButton: PropTypes.bool.isRequired,
  headerNavigationsActions: PropTypes.shape({
    handleNewSection: PropTypes.func.isRequired,
    handleReIndex: PropTypes.func.isRequired,
    handleExpandAll: PropTypes.func.isRequired,
    lmsLink: PropTypes.string.isRequired,
  }).isRequired,
  hasSections: PropTypes.bool.isRequired,
  courseActions: PropTypes.shape({
    deletable: PropTypes.bool.isRequired,
    draggable: PropTypes.bool.isRequired,
    childAddable: PropTypes.bool.isRequired,
    duplicable: PropTypes.bool.isRequired,
  }).isRequired,
  errors: PropTypes.shape({
    outlineIndexApi: PropTypes.shape({
      data: PropTypes.string,
      type: PropTypes.string.isRequired,
    }),
    reindexApi: PropTypes.shape({
      data: PropTypes.string,
      type: PropTypes.string.isRequired,
    }),
    sectionLoadingApi: PropTypes.shape({
      data: PropTypes.string,
      type: PropTypes.string.isRequired,
    }),
    courseLaunchApi: PropTypes.shape({
      data: PropTypes.string,
      type: PropTypes.string.isRequired,
    }),
  }),
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      displayName: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default CourseOutlineHeaderActionsSlot;
