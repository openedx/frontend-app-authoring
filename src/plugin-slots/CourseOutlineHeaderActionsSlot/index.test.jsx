import { shallow } from '@edx/react-unit-test-utils';

import CourseOutlineHeaderActionsSlot from '.';

jest.mock('CourseAuthoring/course-outline/header-navigations/HeaderNavigations', () => 'HeaderNavigations');

jest.mock('@openedx/frontend-plugin-framework', () => ({
  PluginSlot: 'PluginSlot',
}));

// Slot properties that are required only for the default <HeaderNavigation> component
const headerNavProps = {
  isReIndexShow: false,
  isSectionsExpanded: false,
  isDisabledReindexButton: false,
  headerNavigationsActions: {
    handleNewSection: jest.fn(),
    handleReIndex: jest.fn(),
    handleExpandAll: jest.fn(),
    lmsLink: '',
  },
  courseActions: {
    deletable: true,
    draggable: true,
    childAddable: true,
    duplicable: true,
  },
  errors: {},
};

describe('CourseOutlineHeaderActionsSlot', () => {
  beforeEach(() => jest.resetAllMocks());

  test('pluginProps are set correctly', () => {
    const wrapper = shallow(<CourseOutlineHeaderActionsSlot
      sections={[]}
      hasSections={false}
      {...headerNavProps}
    />);
    expect(wrapper.snapshot).toMatchSnapshot();
  });
});
