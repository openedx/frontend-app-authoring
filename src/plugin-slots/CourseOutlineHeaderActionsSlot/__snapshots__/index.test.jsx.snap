// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseOutlineHeaderActionsSlot pluginProps are set correctly 1`] = `
<PluginSlot
  id="course_outline_header_actions_slot"
  pluginProps={
    {
      "courseActions": {
        "childAddable": true,
        "deletable": true,
        "draggable": true,
        "duplicable": true,
      },
      "errors": {},
      "hasSections": false,
      "headerNavigationsActions": {
        "handleExpandAll": [MockFunction],
        "handleNewSection": [MockFunction],
        "handleReIndex": [MockFunction],
        "lmsLink": "",
      },
      "isDisabledReindexButton": false,
      "isReIndexShow": false,
      "isSectionsExpanded": false,
      "sections": [],
    }
  }
>
  <HeaderNavigations
    courseActions={
      {
        "childAddable": true,
        "deletable": true,
        "draggable": true,
        "duplicable": true,
      }
    }
    errors={{}}
    hasSections={false}
    headerNavigationsActions={
      {
        "handleExpandAll": [MockFunction],
        "handleNewSection": [MockFunction],
        "handleReIndex": [MockFunction],
        "lmsLink": "",
      }
    }
    isDisabledReindexButton={false}
    isReIndexShow={false}
    isSectionsExpanded={false}
  />
</PluginSlot>
`;
